beater/beater.go

```diff
@@ -381,31 +415,151 @@ func (s *serverRunner) run() error {
 		}()
 	}
 
-	var sourcemapStore *sourcemap.Store
 	if s.config.RumConfig.Enabled && s.config.RumConfig.SourceMapping.Enabled {
 		store, err := newSourcemapStore(s.beat.Info, s.config.RumConfig.SourceMapping, s.fleetConfig)
 		if err != nil {
 			return err
 		}
-		sourcemapStore = store
+		if s.sourcemapStore == nil {
+			s.sourcemapStore = store
+		} else {
+			*s.sourcemapStore = *store
+		}
+	} else {
+		// TODO: We have to remember to disable the sourcemap if it's
+		// no longer enabled
+		s.sourcemapStore = nil
+	}
+
+	if !s.running {
+		pipeline := pipetool.WithACKer(s.pipeline, s.acker)
+		publisherConfig := &publish.PublisherConfig{
+			Info:      s.beat.Info,
+			Pipeline:  s.config.Pipeline,
+			Namespace: s.namespace,
+		}
+		publisher, err := publish.NewPublisher(pipeline, s.tracer, publisherConfig)
+		if err != nil {
+			return err
+		}
+		s.publisher = publisher
+
+		reporter := publisher.Send
+		s.batchProcessor = &reporterBatchProcessor{reporter}
+		if !s.config.Sampling.KeepUnsampled {
+			// The server has been configured to discard unsampled
+			// transactions. Make sure this is done just before calling
+			// the publisher to avoid affecting aggregations.
+			s.batchProcessor = modelprocessor.Chained{
+				sampling.NewDiscardUnsampledBatchProcessor(),
+				s.batchProcessor,
+			}
+		}
+
+		server, err := newServer(
+			s.logger,
+			s.beat.Info,
+			s.config,
+			s.tracer,
+			reporter,
+			s.sourcemapStore,
+			s.batchProcessor,
+		)
+		if err != nil {
+			return err
+		}
+		s.server = server
+	} else {
+		// TODO: Bit weird to send identical arguments into newServer
+		// and server.configure(). But I want newServer to return a
+		// configured server, and configure() to configure an existing
+		// one.
+		if err := s.server.configure(
+			s.logger,
+			s.beat.Info,
+			s.config,
+			s.tracer,
+			s.publisher.Send,
+			s.sourcemapStore,
+			s.batchProcessor,
+		); err != nil {
+			s.logger.Errorf("failed to configure server: %v", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *serverRunner) String() string {
+	return "APMServer"
+}
+
+// Stop begins a timer, after which the server will shutdown. When the server
+// is being restarted, Stop() and then Start() are called, but we cannot
+// differentiate between that and only receiving Stop().
+// Start a timer, and if we receive a message from the Start() method, do not
+// stop the server.
+// The channel in the timer still needs to be drained.
+func (s *serverRunner) Stop() {
+	t := time.NewTimer(10 * time.Second)
```

Comment:
Reloading is handled by two interfaces:

- There is a struct that "creates" something that can be started and stopped
- There is the thing that starts and stops

I'm assuming we cannot change this interface, so I'm hoping that the start/stop process happens within a reasonable amount of time so we can differentiate between a restart or a shutdown.

I also need to double check that when elastic-agent is shutting down, libbeat only sends a Stop message to the reloadable things it's managing.

@michalpristas can you confirm? When elastic-agent wants to shutdown a managed process, does it simply `Stop()` all of the reloadable items, or does it do something else (like send a unix signal)

Reply:
The interface between Fleet and libbeat is just one interface: [`libbeat/common/reload.Reloadable`](https://github.com/elastic/beats/blob/master/libbeat/common/reload/reload.go).

We are using a layer on top of this which translates reloads into Stop/Start calls: [`libbeat/cfgfile.Reloader`](https://github.com/elastic/beats/blob/master/libbeat/cfgfile/reload.go). We should probably move off this, and implement Reloadable directly.